docs/features/traffic-management/ambassador.md

```diff
@@ -0,0 +1,95 @@
+# Ambassador Edge Stack
+
+[Ambassador Edge Stack](https://www.getambassador.io/products/edge-stack/) provides the functionality you need at the edge your Kubernetes cluster (hence, an "edge stack"). This includes an API gateway, ingress controller, load balancer, developer portal, canary traffic routing and more. It provides a group of CRDs that users can configure to enable different functionalities. 
+
+Argo-Rollouts provides an integration that leverages Ambassador's [canary routing capability](https://www.getambassador.io/docs/latest/topics/using/canary/). This allows the traffic to your application to be gradually incremented while new versions are being deployed.
+
+## How it works
+
+Ambassador Edge Stack provides a resource called `Mapping` that is used to configure how to route traffic to services. Ambassador canary deployment is achieved by creating 2 mappings with the same URL prefix pointing to different services. Consider the following example:
+
+```yaml
+apiVersion: getambassador.io/v2
+kind:  Mapping
+metadata:
+  name: stable-mapping
+spec:
+  prefix: /someapp
+  rewrite: /
+  service: someapp-stable:80
+---
+apiVersion: getambassador.io/v2
+kind:  Mapping
+metadata:
+  name: canary-mapping
+spec:
+  prefix: /someapp
+  rewrite: /
+  service: someapp-canary:80
+  weight: 30
+```
+
+In the example above we are configuring Ambassador to route 30% of the traffic coming from `<public ingress>/someapp` to the service `someapp-canary` and the rest of the traffic will go to the service `someapp-stable`. If users want to gradually increase the traffic to the canary service, they have to update the `canary-mapping` setting the weight to the desired value either manually or automating it somehow. 
+
+With Argo-Rollouts there is no need to create the `canary-mapping`. The process of creating it and gradually updating its weight is fully automated by the Argo-Rollouts controller. The following example shows how to configure the `Rollout` resource to use Ambassador as a traffic router for canary deployments:
+
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Rollout
+...
+spec:
+  strategy:
+    canary:
+      stableService: someapp-stable
+      canaryService: someapp-canary
+      trafficRouting:
+        ambassador:
+          apiVersion: getambassador.io/v2
```

Comment:
For TrafficSplit, Istio VirtualService/DestinationRule CRs, we made apiVersion configuration a controller-level decision with the `--istio-api-version` and `--traffic-split-api-version` flags to the controller. In that way, the rollout controller would choose to work with a single apiVersion of a CRD, and leverage the natural conversion process (e.g. conversion webhooks) to handle any object.

Given TrafficSplit and Istio implementations are deciding apiVersion at the controller level, would it also make sense to do the same for Mapping objects and not expose it here?

Reply:
I saw those flags in the controller when I implemented this attribute. What have driven me to make this as part of the CRD were mainly 2 reasons:
1. It is more flexible to users who might have different mapping versions applied in the cluster to be able to specify which one to use at deployment time. Setting this as argo-rollout flag forces users to use only one mapping version.
2. I found that keeping it at the CRD level allows the Ambassador related code to be mainly centralized in one package making the code cleaner and easier to reason about.

I also understand if you want to keep some level of consistency with other traffic router implementations by adding this as an additional flag.

Please let me know your thoughts.

Reply:
> It is more flexible to users who might have different mapping versions applied in the cluster to be able to specify which one to use at deployment time. Setting this as argo-rollout flag forces users to use only one mapping version.

So My assumption here is that the user's preference of which mapping version to use, doesn't really matter in the end, because they can always retrieve their preferred version.

HPA is a good analogy for this. In kubernetes, you can deploy an `autoscaling/v1` object but still retrieve it from the API as `autoscaling/v2beta1` because kubernetes will convert it on the fly. For example you can run the following two commands to get two versions of the same HPA object

```yaml
kubectl get hpa.v1.autoscaling my-hpa

kubectl hpa.v2beta1.autoscaling my-hpa
```

Is my understanding of CRD conversion incorrect?

Reply:
> Is my understanding of CRD conversion incorrect?

No. You are right that it doesn't necessarily forces users to use only one version as long as they define the latest one to avoid missing fields during the canary mapping generation.

I guess another flexibility problem related to specifying it as flags is that it forces users to reconfigure and redeploy argo-rollouts on the eventual release of a new Ambassador API version. It seems to me that having the ability to specify it at the CRD level provides a better user experience. What do you think?

Reply:
> I guess another flexibility problem related to specifying it as flags is that it forces users to reconfigure and redeploy argo-rollouts on the eventual release of a new Ambassador API version.

This is true, but it's better to put the onus of this on the operator, and not the developers who are actually managing the Rollout objects. In other words, developers will GitOps their Rollout rollout with say a v1 mapping. But when v2 is introduced into the cluster, all of the Rollout objects will then have to be updated to use a v2 mapping. By making it a controller level decision, the end users (developers) do not have to do anything. For this reason I think it's better to remove the apiVersion and have this be a controller level flag.

Reply:
Makes sense. I'll implement the changes and let you know.
Thank you

Reply:
any updates or estimated time for merge?

Reply:
The PR is updated following @jessesuen's suggestion above.
We are now trying to make all checks green again after master rebase.

Reply:
@leoluz amazing thank you!
there is an issue with the sticky session.
the splitting to mapping canary or active is happening before the session affinity action.